Remove DICT_KEYS guard

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #118630
* #118855
* #118535
* #118199
* #118208
* #118003
* #118194
* #117625
* #117983
* #118098
* #117982

This guard was trying to do too much. We replace it for the weaker
DICT_CONST_KEYS, which is only valid when all the values are constants
(i.e. easy to print out).

We also implement an optimisation where, if all the keys to a dictionary
are constant, we generate just one guard of the form
```
list(d.keys()) == ["key1", "key2"]
```

This is the type of guards we used to generate initially before this set
of PRs.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

lezcano committed:
Remove DICT_KEYS guard

This guard was trying to do too much. We replace it for the weaker
DICT_CONST_KEYS, which is only valid when all the values are constants
(i.e. easy to print out).

We also implement an optimisation where, if all the keys to a dictionary
are constant, we generate just one guard of the form
```
list(d.keys()) == ["key1", "key2"]
```

This is the type of guards we used to generate initially before this set
of PRs.

[ghstack-poisoned]

lezcano commented:
The trickiest use of the DICT_KEYS guard was within `optimizer.py`.
There, this guard was used to "manually trace" through `optim.state`.
I have replaced that manual tracing with a call to `VariableBuilder`,
but that currently sets some dupe guards between `optim.state` and
`optim.param_groups["params"]`.

@mlazos do you originally wrote the code
this way to avoid that guard? If so, I can remove it and change it by a
`DICT_ID_KEYS` and still remove `DICT_KEYS`

peterbell10 reviewed (changes requested):
Makes sense, but `test_compiled_optimizers.py` failure looks real.

lezcano commented:
This fixed a graphbreak in asgd and nadam apparently. I have not looked into what happened there really, but I guess that's good.

lezcano commented on torch/_dynamo/guards.py:
```diff
@@ -556,15 +532,15 @@ def NN_MODULE_PARAM_NAMES(self, guard):
 
-    def ODICT_KEYS(self, guard):
-        """OrderedDict keys match"""
+    def DICT_CONST_KEYS(self, guard):
```
I mean, there is a streak of 3 that are in alphabetical order... jajajajaj I'll leave it as is to keep the diff minimal.

lezcano committed:
Update on "Remove DICT_KEYS guard"


This guard was trying to do too much. We replace it for the weaker
DICT_CONST_KEYS, which is only valid when all the values are constants
(i.e. easy to print out).

We also implement an optimisation where, if all the keys to a dictionary
are constant, we generate just one guard of the form
```
list(d.keys()) == ["key1", "key2"]
```

This is the type of guards we used to generate initially before this set
of PRs.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

lezcano committed:
Update on "Remove DICT_KEYS guard"


This guard was trying to do too much. We replace it for the weaker
DICT_CONST_KEYS, which is only valid when all the values are constants
(i.e. easy to print out).

We also implement an optimisation where, if all the keys to a dictionary
are constant, we generate just one guard of the form
```
list(d.keys()) == ["key1", "key2"]
```

This is the type of guards we used to generate initially before this set
of PRs.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

lezcano committed:
Update on "Remove DICT_KEYS guard"


This guard was trying to do too much. We replace it for the weaker
DICT_CONST_KEYS, which is only valid when all the values are constants
(i.e. easy to print out).

We also implement an optimisation where, if all the keys to a dictionary
are constant, we generate just one guard of the form
```
list(d.keys()) == ["key1", "key2"]
```

This is the type of guards we used to generate initially before this set
of PRs.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

lezcano commented:
Triggered a benchmark run to see if there's any perf hit in all the changes in this stack.

peterbell10 commented on test/inductor/test_compiled_optimizers.py:
```diff
@@ -47,12 +47,16 @@ class KernelCounts(NamedTuple):
     "test_adadelta_foreach_weight_decay_maximize_cpu": 12,
     "test_adadelta_foreach_rho_weight_decay_cpu": 12,
     "test_adadelta_foreach_weight_decay_cpu": 12,
+    "test_asgd_recompile_single": 16,
```
NIT: I think this value is defined here

https://github.com/pytorch/pytorch/blob/44632939b5aa0c3e658d807c294df8fb3018fb91/test/inductor/test_compiled_optimizers.py#L320-L323

lezcano committed:
Update on "Remove DICT_KEYS guard"


This guard was trying to do too much. We replace it for the weaker
DICT_CONST_KEYS, which is only valid when all the values are constants
(i.e. easy to print out).

We also implement an optimisation where, if all the keys to a dictionary
are constant, we generate just one guard of the form
```
list(d.keys()) == ["key1", "key2"]
```

This is the type of guards we used to generate initially before this set
of PRs.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

lezcano commented:
The optim changes seem to regress a bit. I'll look into how to get the graphbreak wins from this PR without regressing. I'm merging the rest of the stack, as that seems fine (might need to split the constant dict optimisation into its own PR tho as there are some regressions on gnns otherwise)

lezcano committed:
Update on "Remove DICT_KEYS guard"


This guard was trying to do too much. We replace it for the weaker
DICT_CONST_KEYS, which is only valid when all the values are constants
(i.e. easy to print out).

We also implement an optimisation where, if all the keys to a dictionary
are constant, we generate just one guard of the form
```
list(d.keys()) == ["key1", "key2"]
```

This is the type of guards we used to generate initially before this set
of PRs.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

lezcano committed:
Update on "Remove DICT_KEYS guard"


This guard was trying to do too much. We replace it for the weaker
DICT_CONST_KEYS, which is only valid when all the values are constants
(i.e. easy to print out).

We also implement an optimisation where, if all the keys to a dictionary
are constant, we generate just one guard of the form
```
list(d.keys()) == ["key1", "key2"]
```

This is the type of guards we used to generate initially before this set
of PRs.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]
